File: packages/kits/AriaMcpKit/Tests/AriaMCPTests/AriaV2MemoryListTests.swift
import CryptoKit
import Foundation
import PersistenceKit
import Testing
@testable import AriaMCP

@Suite("ARIA v2 memory-list cursor session")
struct AriaV2MemoryListTests {
    private let estateID = UUID(uuidString: "aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa")!
    private let firstID = UUID(uuidString: "00000000-0000-4000-8000-000000000001")!
    private let secondID = UUID(uuidString: "00000000-0000-4000-8000-000000000002")!
    private let thirdID = UUID(uuidString: "00000000-0000-4000-8000-000000000003")!
    private let now = Date(timeIntervalSince1970: 1_700_000_000)

    private func authorization(_ context: String = "context") -> AriaV2MemoryListAuthorization {
        .init(callerID: "caller", contextID: context, policyVersion: "memory-list-v1")
    }

    private func row(_ id: UUID, subject: String) -> AriaV2MemoryListRow {
        .init(
            memoryID: id,
            ancestryIDs: [], ancestryNames: ["Agentic Memory", "Planning"],
            eligibilityState: "current", visibilityState: "authorized",
            projection: ["subject": .string(subject)])
    }

    private func service(
        provider: FakeSnapshotProvider,
        session: AriaV2MemoryListCursorSession = .init(),
        now: Date? = nil,
        authorization: AriaV2MemoryListAuthorization? = nil
    ) -> AriaV2MemoryListService {
        let instant = now ?? self.now
        return AriaV2MemoryListService(
            provider: provider, cursorSession: session, defaultEstateID: estateID,
            authorization: authorization ?? self.authorization(), now: { instant })
    }

    private func fixture(named name: String) throws -> JSONValue {
        let url = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()
            .appendingPathComponent("../Conformance/\(name)")
            .standardizedFileURL
        return try JSONValue.parse(Data(contentsOf: url))
    }

    @Test func revisionMatchesTheFrozenCrossPortVector() async throws {
        let fixture = try fixture(named: "aria_v2_memory_list_revision_vectors.json")
        let expected = try #require(fixture.objectValue?["expected_sha256"]?.stringValue)
        let canonical = try #require(fixture.objectValue?["canonical_json"]?.stringValue)
        let fixtureDigest = SHA256.hash(data: Data(canonical.utf8))
            .map { String(format: "%02x", $0) }
            .joined()
        #expect(fixtureDigest == expected)

        let parentID = UUID(uuidString: "11111111-1111-4111-8111-111111111111")!
        let wingID = UUID(uuidString: "22222222-2222-4222-8222-222222222222")!
        let first = AriaV2MemoryListRow(
            memoryID: UUID(uuidString: "10000000-0000-4000-8000-000000000000")!,
            ancestryIDs: [parentID, wingID], ancestryNames: ["Agentic Memory", "inbox"],
            eligibilityState: "current", visibilityState: "bulk_exportable",
            projection: ["context": .null, "provenance": .string("agent"),
                         "subject": .null, "weight": .integer(7)])
        let last = AriaV2MemoryListRow(
            memoryID: UUID(uuidString: "f0000000-0000-4000-8000-000000000000")!,
            ancestryIDs: [parentID, wingID], ancestryNames: ["Agentic Memory", "inbox"],
            eligibilityState: "current", visibilityState: "bulk_exportable",
            projection: ["context": .string("planning"), "provenance": .null,
                         "subject": .null, "weight": .integer(8)])
        let provider = FakeSnapshotProvider(snapshot: .init(
            estateID: estateID, authorizationGeneration: "authorization-generation-7",
            rows: [last, first]))
        let result = try await service(
            provider: provider,
            authorization: .init(callerID: "caller-binding-a", contextID: "context-42", policyVersion: "policy-v3")
        ).list(arguments: .object([
            "wing": .string("Agentic Memory"), "filter": .string("missing_subject"),
        ]))
        let revision = try #require(result.objectValue?["structuredContent"]?.objectValue?["data"]?.objectValue?["revision"]?.stringValue)
        #expect(revision == expected)
    }

    @Test func listSortsUUIDBytesAndContinuesOnlyAgainstTheSameFullRevision() async throws {
        let provider = FakeSnapshotProvider(snapshot: .init(
            estateID: estateID, authorizationGeneration: "g1",
            rows: [row(thirdID, subject: "third"), row(firstID, subject: "first"), row(secondID, subject: "second")]))
        let session = AriaV2MemoryListCursorSession()
        let first = try await service(provider: provider, session: session).list(arguments: .object([
            "wing": .string("Agentic Memory"), "limit": .integer(1),
        ]))
        let firstData = try #require(first.objectValue?["structuredContent"]?.objectValue?["data"]?.objectValue)
        #expect(firstData["memories"]?.arrayValue?.first?.objectValue?["memory_id"] == .string(firstID.uuidString.lowercased()))
        #expect(firstData["has_more"] == .bool(true))
        let cursor = try #require(firstData["next_cursor"]?.stringValue)
        let revision = try #require(firstData["revision"]?.stringValue)

        let second = try await service(provider: provider, session: session).list(arguments: .object([
            "wing": .string("Agentic Memory"), "limit": .integer(1), "cursor": .string(cursor),
        ]))
        let secondData = try #require(second.objectValue?["structuredContent"]?.objectValue?["data"]?.objectValue)
        #expect(secondData["memories"]?.arrayValue?.first?.objectValue?["memory_id"] == .string(secondID.uuidString.lowercased()))
        #expect(secondData["revision"] == .string(revision))
        #expect(secondData["has_more"] == .bool(true))
        let finalCursor = try #require(secondData["next_cursor"]?.stringValue)

        let final = try await service(provider: provider, session: session).list(arguments: .object([
            "wing": .string("Agentic Memory"), "limit": .integer(1), "cursor": .string(finalCursor),
        ]))
        let finalData = try #require(final.objectValue?["structuredContent"]?.objectValue?["data"]?.objectValue)
        #expect(finalData["memories"]?.arrayValue?.first?.objectValue?["memory_id"] == .string(thirdID.uuidString.lowercased()))
        #expect(finalData["has_more"] == .bool(false))
        #expect(finalData["next_cursor"] == nil)
    }

    @Test func changedFullStateAndScopeMismatchesRefuseWithoutAPartialPage() async throws {
        let provider = FakeSnapshotProvider(snapshot: .init(
            estateID: estateID, authorizationGeneration: "g1",
            rows: [row(firstID, subject: "first"), row(secondID, subject: "second")]))
        let session = AriaV2MemoryListCursorSession()
        let first = try await service(provider: provider, session: session).list(arguments: .object([
            "wing": .string("Agentic Memory"), "limit": .integer(1),
        ]))
        let cursor = try #require(first.objectValue?["structuredContent"]?.objectValue?["data"]?.objectValue?["next_cursor"]?.stringValue)

        await provider.replace(.init(
            estateID: estateID, authorizationGeneration: "g2",
            rows: [row(firstID, subject: "first changed"), row(secondID, subject: "second")]))
        let stale = try await service(provider: provider, session: session).list(arguments: .object([
            "wing": .string("Agentic Memory"), "limit": .integer(1), "cursor": .string(cursor),
        ]))
        let staleStructured = stale.objectValue?["structuredContent"]?.objectValue
        #expect(staleStructured?["error"]?.objectValue?["code"] == .string("cursor_stale"))
        #expect(staleStructured?["data"] == nil)

        await provider.replace(.init(
            estateID: estateID, authorizationGeneration: "g1",
            rows: [row(firstID, subject: "first"), row(secondID, subject: "second")]))
        let mismatch = try await service(provider: provider, session: session).list(arguments: .object([
            "wing": .string("Other wing"), "limit": .integer(1), "cursor": .string(cursor),
        ]))
        #expect(mismatch.objectValue?["structuredContent"]?.objectValue?["error"]?.objectValue?["code"] == .string("cursor_mismatch"))

        let authorizationMismatch = try await service(
            provider: provider, session: session, authorization: authorization("other-context")
        ).list(arguments: .object([
            "wing": .string("Agentic Memory"), "limit": .integer(1), "cursor": .string(cursor),
        ]))
        #expect(authorizationMismatch.objectValue?["structuredContent"]?.objectValue?["error"]?.objectValue?["code"] == .string("cursor_mismatch"))
    }

    @Test func cursorExpiryAndRetentionCapsUseAbsoluteTTLThenLRU() async throws {
        let session = AriaV2MemoryListCursorSession()
        let scope = AriaV2MemoryListCursorSession.Scope(
            estateID: estateID, wing: "Agentic Memory", room: nil, filter: nil)
        let auth = authorization()
        for index in 0..<33 {
            _ = try await session.retain(
                scope: scope, authorization: auth, revision: "r\(index)",
                lastMemoryID: firstID, now: now.addingTimeInterval(TimeInterval(index)))
        }
        #expect(await session.retainedReferenceCount() == 32)

        for index in 0..<257 {
            _ = try await session.retain(
                scope: scope, authorization: authorization("context-\(index)"), revision: "s\(index)",
                lastMemoryID: secondID, now: now.addingTimeInterval(TimeInterval(100 + index)))
        }
        #expect(await session.retainedReferenceCount() == 256)

        let oversizedScope = AriaV2MemoryListCursorSession.Scope(
            estateID: estateID, wing: "Agentic Memory", room: nil,
            filter: String(repeating: "x", count: AriaV2MemoryListCursorSession.maximumRetainedBytes))
        await #expect(throws: AriaV2MemoryListCursorError.retainedStateLimit) {
            _ = try await session.retain(
                scope: oversizedScope, authorization: auth, revision: "too-large",
                lastMemoryID: thirdID, now: now)
        }

        let provider = FakeSnapshotProvider(snapshot: .init(
            estateID: estateID, authorizationGeneration: "g1",
            rows: [row(firstID, subject: "first"), row(secondID, subject: "second")]))
        let expirySession = AriaV2MemoryListCursorSession()
        let initial = try await service(provider: provider, session: expirySession).list(arguments: .object([
            "wing": .string("Agentic Memory"), "limit": .integer(1),
        ]))
        let cursor = try #require(initial.objectValue?["structuredContent"]?.objectValue?["data"]?.objectValue?["next_cursor"]?.stringValue)
        let expired = try await service(
            provider: provider, session: expirySession,
            now: now.addingTimeInterval(AriaV2MemoryListCursorSession.absoluteTTL)).list(arguments: .object([
                "wing": .string("Agentic Memory"), "limit": .integer(1), "cursor": .string(cursor),
            ]))
        #expect(expired.objectValue?["structuredContent"]?.objectValue?["error"]?.objectValue?["code"] == .string("cursor_expired"))
    }

    @Test func requestIsStrictAndSnapshotLimitFailureHasNoPage() async throws {
        let emptyRoom = try AriaV2MemoryListRequest(arguments: .object([
            "wing": .string("Agentic Memory"), "room": .string(""), "filter": .string("missing_subject"),
        ]))
        #expect(emptyRoom.room == nil)
        #expect(emptyRoom.filter == "missing_subject")
        #expect(throws: JSONRPCError.self) {
            _ = try AriaV2MemoryListRequest(arguments: .object([
                "wing": .string("Agentic Memory"), "unexpected": .bool(true),
            ]))
        }
        #expect(throws: JSONRPCError.self) {
            _ = try AriaV2MemoryListRequest(arguments: .object([
                "wing": .string("Agentic Memory"), "filter": .string("all"),
            ]))
        }
        let provider = FakeSnapshotProvider(error: .rowLimitExceeded(table: "drawers", limit: 250_000))
        let result = try await service(provider: provider).list(arguments: .object([
            "wing": .string("Agentic Memory"),
        ]))
        let structured = result.objectValue?["structuredContent"]?.objectValue
        #expect(structured?["error"]?.objectValue?["code"] == .string("inventory_too_large"))
        #expect(structured?["data"] == nil)
    }
}

private actor FakeSnapshotProvider: AriaV2MemoryListSnapshotProvider {
    private var snapshot: AriaV2MemoryListSnapshot?
    private let error: InventorySnapshotError?

    init(snapshot: AriaV2MemoryListSnapshot) {
        self.snapshot = snapshot
        error = nil
    }

    init(error: InventorySnapshotError) {
        snapshot = nil
        self.error = error
    }

    func replace(_ snapshot: AriaV2MemoryListSnapshot) { self.snapshot = snapshot }

    func immutableAuthorizedSnapshot(
        estateID: UUID,
        wing: String,
        room: String?,
        filter: String?,
        authorization: AriaV2MemoryListAuthorization
    ) async throws -> AriaV2MemoryListSnapshot {
        _ = (estateID, wing, room, filter, authorization)
        if let error { throw error }
        return snapshot!
    }
}

File: packages/kits/AriaMcpKit/rust/tests/aria_v2_memory_list_tests.rs
#![cfg(feature = "aria-v2")]

#[path = "../src/v2/memory_list.rs"]
mod memory_list;

use std::sync::{Arc, Mutex};

use memory_list::{
    MemoryListAuthorization, MemoryListError, MemoryListFilter, MemoryListProjection,
    MemoryListRequest, MemoryListService, MemoryListSnapshot, MemoryListSnapshotProvider,
    MemoryListSnapshotRow, CURSOR_TTL_MILLIS,
};
use serde_json::{json, Map, Value};
use uuid::Uuid;

const NOW: i64 = 1_700_000_000_000;
const ESTATE: &str = "33333333-3333-4333-8333-333333333333";

struct Provider {
    authorization: Mutex<MemoryListAuthorization>,
    snapshot: Mutex<MemoryListSnapshot>,
}

impl MemoryListSnapshotProvider for Provider {
    fn authorize(&self, _: Option<Uuid>) -> Result<MemoryListAuthorization, MemoryListError> {
        Ok(self.authorization.lock().unwrap().clone())
    }

    fn capture_authorized_inventory(
        &self,
        _: &MemoryListAuthorization,
        _: &str,
        _: Option<&str>,
        _: Option<MemoryListFilter>,
    ) -> Result<MemoryListSnapshot, MemoryListError> {
        Ok(self.snapshot.lock().unwrap().clone())
    }

    fn revalidate(&self, _: &MemoryListAuthorization) -> Result<(), MemoryListError> { Ok(()) }
}

impl MemoryListSnapshotProvider for Arc<Provider> {
    fn authorize(&self, requested: Option<Uuid>) -> Result<MemoryListAuthorization, MemoryListError> {
        self.as_ref().authorize(requested)
    }

    fn capture_authorized_inventory(
        &self,
        authorization: &MemoryListAuthorization,
        wing: &str,
        room: Option<&str>,
        filter: Option<MemoryListFilter>,
    ) -> Result<MemoryListSnapshot, MemoryListError> {
        self.as_ref().capture_authorized_inventory(authorization, wing, room, filter)
    }

    fn revalidate(&self, authorization: &MemoryListAuthorization) -> Result<(), MemoryListError> {
        self.as_ref().revalidate(authorization)
    }
}

fn uuid(value: &str) -> Uuid { Uuid::parse_str(value).unwrap() }

fn projection(subject: Option<&str>, provenance: Option<&str>, context: Option<&str>, weight: i64) -> MemoryListProjection {
    Map::from_iter([
        ("subject".to_owned(), subject.map_or(Value::Null, |value| Value::String(value.to_owned()))),
        ("provenance".to_owned(), provenance.map_or(Value::Null, |value| Value::String(value.to_owned()))),
        ("context".to_owned(), context.map_or(Value::Null, |value| Value::String(value.to_owned()))),
        ("weight".to_owned(), Value::Number(weight.into())),
    ])
}

fn row(id: &str, subject: Option<&str>) -> MemoryListSnapshotRow {
    MemoryListSnapshotRow {
        memory_id: uuid(id),
        ancestry_ids: vec![uuid("11111111-1111-4111-8111-111111111111"), uuid("22222222-2222-4222-8222-222222222222")],
        ancestry_names: vec!["Agentic Memory".to_owned(), "inbox".to_owned()],
        eligibility_state: "current".to_owned(),
        visibility_state: "bulk_exportable".to_owned(),
        projection: projection(subject, Some("agent"), None, 7),
    }
}

fn snapshot(rows: Vec<MemoryListSnapshotRow>) -> MemoryListSnapshot {
    MemoryListSnapshot {
        estate_id: uuid(ESTATE), authorization_generation: "generation-1".to_owned(),
        drawer_rows: rows.len(), node_rows: 2, serialized_row_bytes: 128, rows,
    }
}

fn request(cursor: Option<String>) -> MemoryListRequest {
    MemoryListRequest {
        estate_id: None, wing: "Agentic Memory".to_owned(), room: None,
        filter: None, limit: 1, cursor,
    }
}

fn service() -> (MemoryListService<Arc<Provider>>, Arc<Provider>) {
    let provider = Arc::new(Provider {
        authorization: Mutex::new(MemoryListAuthorization {
            caller_binding: "caller-a".to_owned(), context_id: "context-a".to_owned(), policy_version: "policy-v1".to_owned(),
        }),
        snapshot: Mutex::new(snapshot(vec![
            row("f0000000-0000-4000-8000-000000000000", Some("last")),
            row("10000000-0000-4000-8000-000000000000", Some("first")),
            row("80000000-0000-4000-8000-000000000000", Some("middle")),
        ])),
    });
    (MemoryListService::new(Arc::clone(&provider), uuid(ESTATE)), provider)
}

#[test]
fn pages_complete_authorized_state_in_uuid_byte_order() {
    let (service, _) = service();
    let first = service.list(request(None), NOW).unwrap();
    assert_eq!(first.memories[0]["memory_id"], "10000000-0000-4000-8000-000000000000");
    assert!(first.has_more);
    let cursor = first.next_cursor.clone().expect("usable cursor when has_more");

    let second = service.list(request(Some(cursor)), NOW + 1).unwrap();
    assert_eq!(second.memories[0]["memory_id"], "80000000-0000-4000-8000-000000000000");
    assert!(second.has_more);
    assert_eq!(first.revision, second.revision);
}

#[test]
fn stale_expired_and_mismatched_cursors_return_no_page() {
    let (service, provider) = service();
    let first = service.list(request(None), NOW).unwrap();
    let cursor = first.next_cursor.clone().unwrap();

    provider.snapshot.lock().unwrap().rows.push(row("20000000-0000-4000-8000-000000000000", Some("changed")));
    let stale = service.list(request(Some(cursor.clone())), NOW + 1).unwrap_err();
    assert_eq!(stale.code, "cursor_stale");

    let fresh = service.list(request(None), NOW + 2).unwrap();
    let fresh_cursor = fresh.next_cursor.clone().unwrap();
    provider.authorization.lock().unwrap().context_id = "context-b".to_owned();
    let mismatch = service.list(request(Some(fresh_cursor)), NOW + 3).unwrap_err();
    assert_eq!(mismatch.code, "cursor_mismatch");

    provider.authorization.lock().unwrap().context_id = "context-a".to_owned();
    let expiring = service.list(request(None), NOW + 4).unwrap();
    let expired = service.list(request(Some(expiring.next_cursor.unwrap())), NOW + 4 + CURSOR_TTL_MILLIS).unwrap_err();
    assert_eq!(expired.code, "cursor_expired");
}

#[test]
fn decoder_allows_absent_or_missing_subject_filter_and_empty_room_is_nil() {
    let absent = MemoryListRequest::decode(&json!({"wing": "Agentic Memory", "room": ""})).unwrap();
    assert_eq!(absent.room, None);
    assert_eq!(absent.filter, None);
    let filtered = MemoryListRequest::decode(&json!({"wing": "Agentic Memory", "filter": "missing_subject", "limit": 200})).unwrap();
    assert_eq!(filtered.filter, Some(MemoryListFilter::MissingSubject));
    assert!(MemoryListRequest::decode(&json!({"wing":"Agentic Memory","filter":"all"})).is_err());
    assert!(MemoryListRequest::decode(&json!({"wing":"Agentic Memory","extra":true})).is_err());
}

#[test]
fn revision_matches_shared_swift_vector_and_preserves_projection_nulls() {
    let fixture: Value = serde_json::from_str(include_str!("../../Tests/Conformance/aria_v2_memory_list_revision_vectors.json")).unwrap();
    let material = &fixture["revision_material"];
    let authorization = material["authorization"].as_object().unwrap();
    let mut rows = Vec::new();
    for source in material["rows"].as_array().unwrap() {
        let projection = source["projection"].as_object().unwrap().clone();
        rows.push(MemoryListSnapshotRow {
            memory_id: uuid(source["memory_id"].as_str().unwrap()),
            ancestry_ids: source["ancestry_ids"].as_array().unwrap().iter().map(|id| uuid(id.as_str().unwrap())).collect(),
            ancestry_names: source["ancestry_names"].as_array().unwrap().iter().map(|name| name.as_str().unwrap().to_owned()).collect(),
            eligibility_state: source["eligibility"].as_str().unwrap().to_owned(),
            visibility_state: source["visibility"].as_str().unwrap().to_owned(),
            projection,
        });
    }
    let provider = Arc::new(Provider {
        authorization: Mutex::new(MemoryListAuthorization {
            caller_binding: authorization["caller_binding"].as_str().unwrap().to_owned(),
            context_id: authorization["context_id"].as_str().unwrap().to_owned(),
            policy_version: authorization["policy_version"].as_str().unwrap().to_owned(),
        }),
        snapshot: Mutex::new(MemoryListSnapshot {
            estate_id: uuid(material["estate_id"].as_str().unwrap()),
            authorization_generation: authorization["authorization_generation"].as_str().unwrap().to_owned(),
            drawer_rows: rows.len(), node_rows: 2, serialized_row_bytes: 512, rows,
        }),
    });
    let request = MemoryListRequest {
        estate_id: None, wing: material["scope"]["wing"].as_str().unwrap().to_owned(), room: None,
        filter: Some(MemoryListFilter::MissingSubject), limit: 200, cursor: None,
    };
    let page = MemoryListService::new(Arc::clone(&provider), uuid(material["estate_id"].as_str().unwrap()))
        .list(request, NOW).unwrap();
    assert_eq!(page.revision, fixture["expected_sha256"].as_str().unwrap());
    assert_eq!(page.memories[0]["subject"], Value::Null);
    assert_eq!(page.memories[1]["provenance"], Value::Null);
    assert_eq!(page.memories[1]["context"], "planning");
}
